code/espurna/mcp23s08.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+
+MCP23S08 MODULE
+
+Copyright (C) 2020 by Eddi De Pieri <eddi at depieri dot com>
+Copyright (C) 2016-2017 by Xose Pérez <xose dot perez at gmail dot com>
+Copyright (C) 2016 Plamen Kovandjiev <p.kovandiev@kmpelectronics.eu> & Dimitar Antonov <d.antonov@kmpelectronics.eu>
+
+*/
+
+#include "mcp23s08.h"
+
+#if MCP23S08_SUPPORT
+
+#include <bitset>
+
+
+#define MCP23S08_CS 15
+
+#define READ_CMD  0x41
+#define WRITE_CMD 0x40
+
+#define IODIR   0x00
+#define IPOL    0x01
+#define GPINTEN 0x02
+#define DEFVAL  0x03
+#define INTCON  0x04
+#define IOCON   0x05
+#define GPPU    0x06
+#define INTF    0x07
+#define INTCAP  0x08
+#define GPIO    0x09
+#define OLAT    0x0A
+
+/**
+ * @brief Relay pins.
+ */
+const uint8_t RELAY_PINS[DUMMY_RELAY_COUNT] =
+{ MCP23S08_REL1PIN, MCP23S08_REL2PIN, MCP23S08_REL3PIN, MCP23S08_REL4PIN };
+
+/**
+ * @brief Input pins.
+ */
+const int OPTOIN_PINS[MCP23S08_OPTOIN_COUNT] =
+{ MCP23S08_IN1PIN, MCP23S08_IN2PIN, MCP23S08_IN3PIN, MCP23S08_IN4PIN };
+
+uint8_t  _expTxData[16]  __attribute__((aligned(4)));
+uint8_t  _expRxData[16]  __attribute__((aligned(4))); 
+
+
+McpGpioPin::McpGpioPin(unsigned char pin) :
+    BasePin(pin)
+{}
+
+inline void McpGpioPin::pinMode(int8_t mode) {
+    ::MCP23S08SetDirection(this->pin, mode);
+}
+
+inline void McpGpioPin::digitalWrite(int8_t val) {
+    ::MCP23S08SetRelayState(this->pin, val);
+}
+
+inline int McpGpioPin::digitalRead() {
+    return ::MCP23S08GetOptoInState(this->pin);
+}
+
+void MCP23S08Setup()
+{
+    DEBUG_MSG_P(PSTR("[MCP23S08] Initialize SPI bus\n"));
+    // Expander settings.
+    SPI.begin();
+    SPI.setHwCs(MCP23S08_CS);
+    SPI.setFrequency(1000000);
```

Comment:
Slightly confused about this.
Per http://ww1.microchip.com/downloads/en/DeviceDoc/MCP23008-MCP23S08-Data-Sheet-20001919F.pdf
SPI bus frequency is said to be 10MHz, but here it is 1MHz. Either SPI class does not care or there is some weird thing SPI does that 'sets' frequency for both.

Reply:
Also regarding SPI, I think we can just call begin(). Check out https://github.com/esp8266/Arduino/blob/cc1cc0b2ce4825057f6a8bb0c6b1b82e17be313e/libraries/SPI/SPI.cpp#L65-L87

setHwCs accepts boolean, so pin 15 cannot be changed
I wonder if digitalWrite() calls are also not needed and handled by the SPI library, as it *seems* those are meant to signal SPI busy / not busy, which SPI.cpp kind of does through some weird register calls.